lib/modules/ai_bot/personas/persona.rb

```diff
@@ -0,0 +1,119 @@
+          result
+        end
+
+        def render_system_prompt(topic: nil, render_function_instructions: true)
+          substitutions = {
+            site_url: Discourse.base_url,
+            site_title: SiteSetting.title,
+            site_description: SiteSetting.site_description,
+            time: Time.zone.now,
+            commands: render_commands(render_function_instructions: render_function_instructions),
+          }
+
+          substitutions[:participants] = topic.allowed_users.map(&:username).join(", ") if topic
+
+          system_prompt.gsub(/\{(\w+)\}/) do |match|
+            found = substitutions[match[1..-2].to_sym]
+            found.nil? ? match : found.to_s
+          end
+        end
+
+        def available_commands
+          return @available_commands if @available_commands
+
+          @available_commands = all_available_commands.filter { |cmd| commands.include?(cmd) }
+        end
+
+        def available_functions
+          # note if defined? can be a problem in test
+          # this can never be nil so it is safe
+          return @available_functions if @available_functions
+
+          functions = []
+
+          functions =
+            available_commands.map do |command|
+              function =
+                DiscourseAi::Inference::Function.new(name: command.name, description: command.desc)
+              command.parameters.each do |parameter|
+                function.add_parameter(
+                  name: parameter.name,
+                  type: parameter.type,
+                  description: parameter.description,
+                  required: parameter.required,
+                  enum: parameter.enum,
+                )
+              end
+              function
+            end
+
+          @available_functions = functions
+        end
+
+        def function_list
+          return @function_list if @function_list
+
+          @function_list = DiscourseAi::Inference::FunctionList.new
+          available_functions.each { |function| @function_list << function }
+          @function_list
+        end
+
+        def all_available_commands
+          return @cmds if @cmds
+
+          all_commands =
+            [
+              Commands::CategoriesCommand,
+              Commands::TimeCommand,
+              Commands::SearchCommand,
+              Commands::SummarizeCommand,
+              Commands::ReadCommand,
+              Commands::SettingContextCommand,
+            ].tap do |cmds|
+              cmds << Commands::TagsCommand if SiteSetting.tagging_enabled
+              cmds << Commands::ImageCommand if SiteSetting.ai_stability_api_key.present?
+              if SiteSetting.ai_google_custom_search_api_key.present? &&
+                   SiteSetting.ai_google_custom_search_cx.present?
+                cmds << Commands::GoogleCommand
+              end
+            end
+
+          allowed_commands = SiteSetting.ai_bot_enabled_chat_commands.split("|")
```

Comment:
For this you can also use the `_map` magic, e.g. `SiteSetting.ai_bot_enabled_chat_commands_map`

Reply:
undefined method for that, I guess something not right about the site setting... technically though this is a list not a map?

Reply:
Ah @SamSaffron it's because if you don't define a type for the `list` site settings it returns nil from the type supervisor, and I only add the `_map` method for types `simple` and `compact:

https://github.com/discourse/discourse/blob/d3a5156e6635d6c33c42d12f4dc205646907cddd/lib/site_setting_extension.rb#L530-L533

Also I chose `_map` for the suffix because of ruby + js `.map()` returning an array. Maybe could have been `_list` in retrospect! :)

Reply:
Added here:

https://github.com/discourse/discourse/pull/23331